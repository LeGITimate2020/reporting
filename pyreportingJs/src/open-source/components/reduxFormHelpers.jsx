import React from 'react';
import { Field } from 'redux-form';
import _ from 'lodash';
import DatePicker from 'react-datepicker';
import moment from 'moment';
import LabelFormEditor from './LabelFormEditor';

const renderFieldWithInlineErrors = (field) => {
  const {
    input,
    label,
    type,
    placeholder,
    className,
    onKeyDown,
    tabIndex,
    meta: { touched, error },
    disabled,
    microcopy,
  } = field;

  if (type === 'checkbox') {
    return (
      <div className="form-group form-inline form-check">
        <label htmlFor={input.name} className="form-check-label">
          <input
            {...input}
            type={type}
            placeholder={placeholder}
            className="form-check-input"
            tabIndex={tabIndex}
            disabled={disabled}
          />
          {label}
          {microcopy && (<span className="microcopy">{microcopy}</span>)}
        </label>
        {touched && error && <div style={{ color: 'red', fontWeight: 'bold' }}>{error}</div>}
      </div>
    );
  }
  return (
    <div className="form-group">
      {label && (
        <label htmlFor={input.name}>
          {label}
          {microcopy && (<span className="microcopy">{microcopy}</span>)}
        </label>
      )}
      <div>
        <input
          {...input}
          onKeyDown={onKeyDown}
          type={type}
          placeholder={placeholder}
          className={`${className} form-control`}
          tabIndex={tabIndex}
          disabled={disabled}
        />
        {touched && error && <span style={{ color: 'red', fontWeight: 'bold' }}>{error}</span>}
      </div>
    </div>
  );
};

export const StandardInput = ({
  inputName = '',
  friendlyName = '',
  inputType = 'text',
  passThrough = {},
}) => (
  <Field
    component={renderFieldWithInlineErrors}
    name={inputName}
    label={friendlyName}
    type={inputType}
    className="form-control"
    {...passThrough}
  />
);

export const InlineInput = ({
  inputName = '',
  placeholder = '',
  inputType = 'text',
  customControls = null,
  passThrough = {},
}) => {
  _.extend(passThrough, { placeholder });
  return (
    <div className="form-group">
      <Field
        {...passThrough}
        component={renderFieldWithInlineErrors}
        name={inputName}
        type={inputType}
        className={`form-control input-sm ${_.get(passThrough, 'className')}`}
      />
      {customControls}
    </div>
  );
};


const textAreaWithErrors = (field) => {
  const {
    meta: { touched, error },
    input,
    ...passThrough
  } = field;
  return (
    <div>
      <textarea {...input} {...passThrough} />
      {touched && error && <div style={{ color: 'red', fontWeight: 'bold' }}>{error}</div>}
    </div>
  );
};


export const StandardTextArea = ({
  inputName = '',
  friendlyName = '',
  customControls = null,
  // TODO: consolidate the microcopy API. when the Field component is renderFieldWithInlineErrors
  // it's in passThrough, for others, 'microcopy' is in the root options map.
  microcopy = '',
  passThrough = {},
}) => {
  return (
    <div className="form-group">
      <label htmlFor={inputName}>
        {friendlyName}
        {microcopy && (<span className="microcopy">{microcopy}</span>)}
      </label>
      <Field component={textAreaWithErrors} name={inputName} className="form-control" {...passThrough} />
      { customControls }
    </div>
  );
};

export const StandardSelect = ({
  selectName = '',
  friendlyName = '',
  optionValues = [],
  optionFriendlyNames = [],
  onChange = null,
  microcopy = '',
  allowBlank = true,
  passThrough = {},
}) => (
  <div className="form-group">
    <label htmlFor={selectName}>
      {friendlyName}
      {microcopy && (<span className="microcopy">{microcopy}</span>)}
    </label>
    <div>
      <Field name={selectName} component="select" className="form-control" onChange={onChange} {...passThrough}>
        { allowBlank ? (<option />) : '' }
        {
          optionFriendlyNames.map((optionFriendlyName, index) => <option key={index} value={optionValues[index]}>{optionFriendlyName}</option>)
        }
      </Field>
    </div>
  </div>
);

export const InlineSelect = ({
  selectName = '',
  placeholder = '',
  label = '',
  optionValues = [],
  optionFriendlyNames = [],
  onChange = null,
  allowBlank = true,
  passThrough = {},
}) => {
  _.extend(passThrough, { placeholder });
  return (
    <div className="form-group">
      <div>
        {label && (
          <label htmlFor={selectName}>{label}: </label>
        )}
        <Field
          {...passThrough}
          name={selectName}
          component="select"
          onChange={onChange}
          className={`form-control input-sm ${_.get(passThrough, 'className')}`}
        >
          { allowBlank ? (<option />) : '' }
          {
            optionFriendlyNames.map((optionFriendlyName, index) => <option key={index} value={optionValues[index]}>{optionFriendlyName}</option>)
          }
        </Field>
      </div>
    </div>
  );
};

const renderStandardLabels = (field) => {
  const { input, suggestions } = field;
  return (
    <div className="form-group">
      <label htmlFor={input.name}>Labels:</label>

      <LabelFormEditor
        name={input.name}
        labels={input.value || []}
        onChange={input.onChange}
        suggestions={suggestions}
      />
      <br />
    </div>
  );
};

export const StandardLabels = ({ name, suggestions }) => {
  return (
    <Field
      name={name}
      component={renderStandardLabels}
      suggestions={suggestions}
    />
  );
};

/**
 *
 * @param props is a superset of the field object from Redux-form
 */
const renderDatePicker = (props) => {
  const { input, label } = props;
  let { dateFormat } = props;
  dateFormat = dateFormat || 'MM/DD/YYYY';
  // the 3rd-party DatePicker requires the value prop to be a date string, not a
  // moment or Date object.
  if (typeof (input.value) !== 'string') {
    input.value = moment(input.value).format(dateFormat);
  }
  return (
    <div className="form-group">
      <label htmlFor={input.name}>{label || 'Date'}:</label>
      <div>
        <DatePicker
          maxDate={moment()}
          peekNextMonth
          showMonthDropdown
          showYearDropdown
          dropdownMode="select"
          {...input}
          {...props}
          dateFormat={dateFormat}
          selected={input.value ? moment(input.value) : null}
          className="form-control"
        />
      </div>
    </div>
  );
};

/**
 *
 * @param props is a superset of the field object from Redux-form
 */
const renderInlineDatePicker = (props) => {
  const { input, className, placeholder } = props;
  let { dateFormat } = props;
  dateFormat = dateFormat || 'MM/DD/YYYY';
  // the 3rd-party DatePicker requires the value prop to be a date string, not a
  // moment or Date object.
  if (typeof (input.value) !== 'string') {
    input.value = moment(input.value).format(dateFormat);
  }
  return (
    <div className="form-group">
      <DatePicker
        maxDate={moment()}
        peekNextMonth
        showMonthDropdown
        showYearDropdown
        dropdownMode="select"
        {...input}
        {...props}
        placeholderText={placeholder}
        dateFormat={dateFormat}
        selected={input.value ? moment(input.value) : null}
        className={`form-control input-sm ${className}`}
      />
    </div>
  );
};

export const StandardDatePicker = ({ name, label, onChange }) => {
  return (
    <Field
      name={name}
      component={renderDatePicker}
      label={label}
      onChange={onChange}
    />
  );
};

export const InlineDatePicker = ({
  name, placeholder, onBlur, maxDate = null,
} = {}) => {
  return (
    <Field
      name={name}
      component={renderInlineDatePicker}
      placeholder={placeholder}
      className="input-narrow"
      onBlur={onBlur}
      maxDate={maxDate}
    />
  );
};


export const RangeSlider = (props) => {
  const {
    input, label, min, max, step,
  } = props;
  return (
    <div>
      <div className="form-group">
        {label && (<label htmlFor={input.name}>{label}:</label>)}
        <div>
          <input type="range" min={min} max={max} step={step} {...input} />
          <span>{input.value}%</span>
        </div>
      </div>
    </div>
  );
};

export const DropDownList = ({ name, friendlyName, values }) => {
  return (
    <StandardSelect
      selectName={name}
      friendlyName={friendlyName || name}
      optionValues={values}
      optionFriendlyNames={values}
      allowBlank
    />
  );
};
