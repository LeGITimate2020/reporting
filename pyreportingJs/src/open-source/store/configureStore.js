import { createStore, applyMiddleware, compose } from 'redux';
import thunkMiddleware from 'redux-thunk';
import reducer from './reducers';

const middleware = [
  thunkMiddleware,
];

/**
 * Initialize a Redux store with initial state, middleware, and Redux Dev Tools
 */
const configureStore = (initialState) => {
  // { serialize: true } enables ES6 types like Map() and Set() to show up in Redux Dev Tools
  const devToolsConfig = { serialize: true };
  return createStore(
    reducer,
    initialState,
    compose(
      applyMiddleware(...middleware),
      window.devToolsExtension ? window.devToolsExtension(devToolsConfig) : f => f
    )
  );
};

export default configureStore;
