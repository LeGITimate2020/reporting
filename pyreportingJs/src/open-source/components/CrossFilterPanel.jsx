import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { reduxForm, formValueSelector } from 'redux-form';
import { connect } from 'react-redux';
import * as d3 from 'd3';
import crossfilter from 'crossfilter2';
import dc from 'dc';
import { ScatterPlot } from '../../react-dc/react-dc';
import { DropDownList } from './reduxFormHelpers';
import '../../react-dc/react-dc.css';


/**
 * This component should accept data and then be the hub through which you can add multiple related graphs. Currently
 * the only graph implemented is the ScatterPlot.
 */
class CrossFilterPanel extends Component {
  static propTypes = {
    /**
     * Sample:
     *  [
     *    {date: "2011-11-14T16:17:54Z", quantity: 2, total: 190, tip: 100, type: "tab", productIDs:["001"]},
     *    ...
     *  ]
     * More detail here: https://github.com/crossfilter/crossfilter/wiki/API-Reference
     */
    data: PropTypes.arrayOf(PropTypes.object).isRequired,
  };

  constructor(props) {
    super(props);
    this.state = this._initializeCrossFilter(props.data);
  }

  _initializeCrossFilter = (data) => {
    // TODO: do validation/sanity checks on the data?

    // these SHOULD NOT live in this.state because they are rich objects and we will be mutating them directly
    this.cfData = crossfilter(data);
    this.dimensions = {};
    this.groups = {};

    const columns = this._getColumnNamesFromData(data);

    // these SHOULD all live in this.state because they are simple objects that will only be read/re-created
    return {
      columns,
      dimensionNames: [],
      charts: [],
    };
  };

  _getColumnNamesFromData = (data) => {
    if (Array.isArray(data) && data.length > 0) {
      return Object.keys(data[0]);
    }
    return [];
  };

  renderCFGroup = (dimensionName) => {
    return this.groups[dimensionName].all().map(item => (
      <div key={item.key}>
        {item.key} = {item.value}
        <button onClick={() => { this.filter(dimensionName, item.key); }}>Filter {dimensionName} to {item.key}</button>
      </div>
    ));
  };

  filter = (dimension, value) => {
    this.dimensions[dimension].filter(value);
    this.forceUpdate();
  };

  createDimensionAndGroup = (dimensionName, dimensionFn) => {
    if (dimensionName in this.dimensions) {
      console.log(`${dimensionName} already exists. Skipping creating a new one.`);
      return;
    }

    const _dimFn = typeof dimensionFn !== 'function' ? d => d[dimensionName] : dimensionFn;

    // create new crossfilter dimension & group
    const dimension = this.cfData.dimension(_dimFn);
    const group = dimension.group();

    // store these in instance Objects for easy access throughout this component
    this.dimensions[dimensionName] = dimension;
    this.groups[dimensionName] = group;

    this.setState({
      dimensionNames: [...this.state.dimensionNames, dimensionName],
    });

    return [dimension, group];
  };

  renderDimension = (dimensionName) => {
    return (
      <div key={dimensionName}>
        <h4>{dimensionName}</h4>
        {this.renderCFGroup(dimensionName)}
      </div>
    );
  };

  renderDimensions = () => {
    return Object.keys(this.dimensions).map(d => this.renderDimension(d));
  };

  render() {
    const { columns } = this.state;
    const { column1, column2, title } = this.props;
    return (
      <div className="container-fluid">
        <div className="row">
          <h2>{title}</h2>
          <div className="col-lg-2">
            xAxis
            <DropDownList name="column1" values={columns} />
          </div>
          <div className="col-lg-2">
            yAxis
            <DropDownList name="column2" values={columns} />
          </div>
          <div className="col-lg-8">&nbsp;</div>
        </div>
        <div className="row">
          <button className="btn btn-primary" onClick={() => { this.createScatter(column1, column2); }}>
            Create scatter: ({column1}, {column2})
          </button>

          {this.renderScatters()}
        </div>

      </div>
    );
  }

  /**
   *
   * @param column1 - xAxis column name
   * @param column2 - yAxis column name
   * @param props - any additional props you'd like to passthrough to the ScatterPlot component
   */
  createScatter = (column1, column2, props = {}) => {
    const dimensionName = `scatter-${column1}-${column2}`;
    const [dimension, group] = this.createDimensionAndGroup(dimensionName, r => [r[column1], r[column2]]);
    const [xMin, xMax] = d3.extent(this.props.data, d => d[column1]);
    const [yMin, yMax] = d3.extent(this.props.data, d => d[column2]);
    const scaleFactor = 0.1;

    const chart = {
      // this type is some term we made up so that we could do some switch/case rendering
      type: 'scatter',

      props: {
        key: dimensionName,

        // these props are props from react-dc
        dimension,
        group,
        x: d3.scaleLinear().domain([xMin - (Math.abs(xMin) * scaleFactor), xMax * (1 + scaleFactor)]),
        y: d3.scaleLinear().domain([yMin - (Math.abs(yMin) * scaleFactor), yMax * (1 + scaleFactor)]),
        xAxisLabel: column1,
        yAxisLabel: column2,
        height: 500,
        width: 500,

        // allow caller to inject any other props, or override props defined above (e.g. height/width)
        ...props,
      },
    };

    this.setState(state => ({ charts: [...state.charts, chart] }));
  };

  renderScatters = () => {
    // Find all scatter charts
    const scatterCharts = this.state.charts.filter(chart => chart.type === 'scatter');

    return (
      <div>
        { scatterCharts.map(chart => (<ScatterPlot {...chart.props} />)) }
      </div>
    );
  };

  clearAllFilters = () => {
    for (const [dimensionName, dimension] of Object.entries(this.dimensions)) {
      dimension.filterAll();
    }
    dc.renderAll();
    this.forceUpdate();
  };
}

const formName = 'crossFilterPanel';
const ReduxForm = reduxForm({
  form: formName,
  enableReinitialize: true,
})(CrossFilterPanel);

const selector = formValueSelector(formName);

const mapStateToProps = (state, ownProps) => {
  const initialValues = {};
  return {
    initialValues,
    column1: selector(state, 'column1'),
    column2: selector(state, 'column2'),
    ...ownProps,
  };
};

export default connect(mapStateToProps)(ReduxForm);
