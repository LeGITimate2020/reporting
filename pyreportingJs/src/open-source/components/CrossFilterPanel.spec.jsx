import { shallow } from 'enzyme';
import React from 'react';
import configureStore from 'redux-mock-store';
import CrossFilterPanel from './CrossFilterPanel';
import { DropDownList } from './reduxFormHelpers';

describe('CrossFilterPanel.jsx', () => {
  const data = [
    {
      date: '2011-11-14T16:17:54Z', quantity: 2, total: 190, tip: 100, type: 'tab', productIDs: ['001'],
    },
    {
      date: '2011-11-14T16:20:19Z', quantity: 2, total: 190, tip: 100, type: 'tab', productIDs: ['001', '005'],
    },
    {
      date: '2011-11-14T16:28:54Z', quantity: 1, total: 300, tip: 200, type: 'visa', productIDs: ['004', '005'],
    },
    {
      date: '2011-11-14T16:30:43Z', quantity: 2, total: 90, tip: 0, type: 'tab', productIDs: ['001', '002'],
    },
    {
      date: '2011-11-14T16:48:46Z', quantity: 2, total: 90, tip: 0, type: 'tab', productIDs: ['005'],
    },
    {
      date: '2011-11-14T16:53:41Z', quantity: 2, total: 90, tip: 0, type: 'tab', productIDs: ['001', '004', '005'],
    },
    {
      date: '2011-11-14T16:54:06Z', quantity: 1, total: 100, tip: 0, type: 'cash', productIDs: ['001', '002', '003', '004', '005'],
    },
    {
      date: '2011-11-14T16:58:03Z', quantity: 2, total: 90, tip: 0, type: 'tab', productIDs: ['001'],
    },
    {
      date: '2011-11-14T17:07:21Z', quantity: 2, total: 90, tip: 0, type: 'tab', productIDs: ['004', '005'],
    },
    {
      date: '2011-11-14T17:22:59Z', quantity: 2, total: 90, tip: 0, type: 'tab', productIDs: ['001', '002', '004', '005'],
    },
    {
      date: '2011-11-14T17:25:45Z', quantity: 2, total: 200, tip: 0, type: 'cash', productIDs: ['002'],
    },
    {
      date: '2011-11-14T17:29:52Z', quantity: 1, total: 200, tip: 100, type: 'visa', productIDs: ['004'],
    },
  ];
  const store = configureStore()();
  const props = {
    data,
    store,
  };

  const diveToComponent = (wrapper) => {
    return wrapper
      .dive()
      .dive()
      .dive()
      .dive();
  };

  it('renders without crashing', () => {
    diveToComponent(shallow(<CrossFilterPanel {...props} />));
  });

  it('Return CrossFilterPanel with correct props', () => {
    const wrapper = diveToComponent(shallow(<CrossFilterPanel {...props} />));

    // Should be 2 DropDownList components (x and y axes)
    expect(wrapper.find(DropDownList).length).toBe(2);
  });

  it('Matches snapshot', () => {
    const wrapper = diveToComponent(shallow(<CrossFilterPanel {...props} />));
    expect(wrapper).toMatchSnapshot();
  });
});
