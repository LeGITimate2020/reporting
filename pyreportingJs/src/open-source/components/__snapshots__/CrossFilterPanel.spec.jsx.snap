// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrossFilterPanel.jsx Matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CrossFilterPanel
    anyTouched={false}
    array={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "removeAll": [Function],
        "shift": [Function],
        "splice": [Function],
        "swap": [Function],
        "unshift": [Function],
      }
    }
    asyncValidate={[Function]}
    asyncValidating={false}
    autofill={[Function]}
    blur={[Function]}
    change={[Function]}
    clearAsyncError={[Function]}
    clearFields={[Function]}
    clearSubmit={[Function]}
    clearSubmitErrors={[Function]}
    column1={undefined}
    column2={undefined}
    data={
      Array [
        Object {
          "date": "2011-11-14T16:17:54Z",
          "productIDs": Array [
            "001",
          ],
          "quantity": 2,
          "tip": 100,
          "total": 190,
          "type": "tab",
        },
        Object {
          "date": "2011-11-14T16:20:19Z",
          "productIDs": Array [
            "001",
            "005",
          ],
          "quantity": 2,
          "tip": 100,
          "total": 190,
          "type": "tab",
        },
        Object {
          "date": "2011-11-14T16:28:54Z",
          "productIDs": Array [
            "004",
            "005",
          ],
          "quantity": 1,
          "tip": 200,
          "total": 300,
          "type": "visa",
        },
        Object {
          "date": "2011-11-14T16:30:43Z",
          "productIDs": Array [
            "001",
            "002",
          ],
          "quantity": 2,
          "tip": 0,
          "total": 90,
          "type": "tab",
        },
        Object {
          "date": "2011-11-14T16:48:46Z",
          "productIDs": Array [
            "005",
          ],
          "quantity": 2,
          "tip": 0,
          "total": 90,
          "type": "tab",
        },
        Object {
          "date": "2011-11-14T16:53:41Z",
          "productIDs": Array [
            "001",
            "004",
            "005",
          ],
          "quantity": 2,
          "tip": 0,
          "total": 90,
          "type": "tab",
        },
        Object {
          "date": "2011-11-14T16:54:06Z",
          "productIDs": Array [
            "001",
            "002",
            "003",
            "004",
            "005",
          ],
          "quantity": 1,
          "tip": 0,
          "total": 100,
          "type": "cash",
        },
        Object {
          "date": "2011-11-14T16:58:03Z",
          "productIDs": Array [
            "001",
          ],
          "quantity": 2,
          "tip": 0,
          "total": 90,
          "type": "tab",
        },
        Object {
          "date": "2011-11-14T17:07:21Z",
          "productIDs": Array [
            "004",
            "005",
          ],
          "quantity": 2,
          "tip": 0,
          "total": 90,
          "type": "tab",
        },
        Object {
          "date": "2011-11-14T17:22:59Z",
          "productIDs": Array [
            "001",
            "002",
            "004",
            "005",
          ],
          "quantity": 2,
          "tip": 0,
          "total": 90,
          "type": "tab",
        },
        Object {
          "date": "2011-11-14T17:25:45Z",
          "productIDs": Array [
            "002",
          ],
          "quantity": 2,
          "tip": 0,
          "total": 200,
          "type": "cash",
        },
        Object {
          "date": "2011-11-14T17:29:52Z",
          "productIDs": Array [
            "004",
          ],
          "quantity": 1,
          "tip": 100,
          "total": 200,
          "type": "visa",
        },
      ]
    }
    destroy={[Function]}
    dirty={false}
    dispatch={[Function]}
    error={undefined}
    form="crossFilterPanel"
    handleSubmit={[Function]}
    initialValues={Object {}}
    initialize={[Function]}
    initialized={false}
    invalid={false}
    pristine={true}
    pure={true}
    reset={[Function]}
    resetSection={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    submit={[Function]}
    submitFailed={false}
    submitSucceeded={false}
    submitting={false}
    touch={[Function]}
    triggerSubmit={undefined}
    untouch={[Function]}
    valid={true}
    warning={undefined}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="row"
        >
          <h2 />
          <div
            className="col-lg-2"
          >
            xAxis
            <Unknown
              name="column1"
              values={
                Array [
                  "date",
                  "quantity",
                  "total",
                  "tip",
                  "type",
                  "productIDs",
                ]
              }
            />
          </div>
          <div
            className="col-lg-2"
          >
            yAxis
            <Unknown
              name="column2"
              values={
                Array [
                  "date",
                  "quantity",
                  "total",
                  "tip",
                  "type",
                  "productIDs",
                ]
              }
            />
          </div>
          <div
            className="col-lg-8"
          >
             
          </div>
        </div>,
        <div
          className="row"
        >
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Create scatter: (
            , 
            )
          </button>
          <div />
        </div>,
      ],
      "className": "container-fluid",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2 />,
            <div
              className="col-lg-2"
            >
              xAxis
              <Unknown
                name="column1"
                values={
                  Array [
                    "date",
                    "quantity",
                    "total",
                    "tip",
                    "type",
                    "productIDs",
                  ]
                }
              />
            </div>,
            <div
              className="col-lg-2"
            >
              yAxis
              <Unknown
                name="column2"
                values={
                  Array [
                    "date",
                    "quantity",
                    "total",
                    "tip",
                    "type",
                    "productIDs",
                  ]
                }
              />
            </div>,
            <div
              className="col-lg-8"
            >
               
            </div>,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "xAxis",
                <Unknown
                  name="column1"
                  values={
                    Array [
                      "date",
                      "quantity",
                      "total",
                      "tip",
                      "type",
                      "productIDs",
                    ]
                  }
                />,
              ],
              "className": "col-lg-2",
            },
            "ref": null,
            "rendered": Array [
              "xAxis",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "name": "column1",
                  "values": Array [
                    "date",
                    "quantity",
                    "total",
                    "tip",
                    "type",
                    "productIDs",
                  ],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "yAxis",
                <Unknown
                  name="column2"
                  values={
                    Array [
                      "date",
                      "quantity",
                      "total",
                      "tip",
                      "type",
                      "productIDs",
                    ]
                  }
                />,
              ],
              "className": "col-lg-2",
            },
            "ref": null,
            "rendered": Array [
              "yAxis",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "name": "column2",
                  "values": Array [
                    "date",
                    "quantity",
                    "total",
                    "tip",
                    "type",
                    "productIDs",
                  ],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": " ",
              "className": "col-lg-8",
            },
            "ref": null,
            "rendered": " ",
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="btn btn-primary"
              onClick={[Function]}
            >
              Create scatter: (
              , 
              )
            </button>,
            <div />,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Create scatter: (",
                undefined,
                ", ",
                undefined,
                ")",
              ],
              "className": "btn btn-primary",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              "Create scatter: (",
              undefined,
              ", ",
              undefined,
              ")",
            ],
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
            },
            "ref": null,
            "rendered": Array [],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="row"
          >
            <h2 />
            <div
              className="col-lg-2"
            >
              xAxis
              <Unknown
                name="column1"
                values={
                  Array [
                    "date",
                    "quantity",
                    "total",
                    "tip",
                    "type",
                    "productIDs",
                  ]
                }
              />
            </div>
            <div
              className="col-lg-2"
            >
              yAxis
              <Unknown
                name="column2"
                values={
                  Array [
                    "date",
                    "quantity",
                    "total",
                    "tip",
                    "type",
                    "productIDs",
                  ]
                }
              />
            </div>
            <div
              className="col-lg-8"
            >
               
            </div>
          </div>,
          <div
            className="row"
          >
            <button
              className="btn btn-primary"
              onClick={[Function]}
            >
              Create scatter: (
              , 
              )
            </button>
            <div />
          </div>,
        ],
        "className": "container-fluid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2 />,
              <div
                className="col-lg-2"
              >
                xAxis
                <Unknown
                  name="column1"
                  values={
                    Array [
                      "date",
                      "quantity",
                      "total",
                      "tip",
                      "type",
                      "productIDs",
                    ]
                  }
                />
              </div>,
              <div
                className="col-lg-2"
              >
                yAxis
                <Unknown
                  name="column2"
                  values={
                    Array [
                      "date",
                      "quantity",
                      "total",
                      "tip",
                      "type",
                      "productIDs",
                    ]
                  }
                />
              </div>,
              <div
                className="col-lg-8"
              >
                 
              </div>,
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "xAxis",
                  <Unknown
                    name="column1"
                    values={
                      Array [
                        "date",
                        "quantity",
                        "total",
                        "tip",
                        "type",
                        "productIDs",
                      ]
                    }
                  />,
                ],
                "className": "col-lg-2",
              },
              "ref": null,
              "rendered": Array [
                "xAxis",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "name": "column1",
                    "values": Array [
                      "date",
                      "quantity",
                      "total",
                      "tip",
                      "type",
                      "productIDs",
                    ],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "yAxis",
                  <Unknown
                    name="column2"
                    values={
                      Array [
                        "date",
                        "quantity",
                        "total",
                        "tip",
                        "type",
                        "productIDs",
                      ]
                    }
                  />,
                ],
                "className": "col-lg-2",
              },
              "ref": null,
              "rendered": Array [
                "yAxis",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "name": "column2",
                    "values": Array [
                      "date",
                      "quantity",
                      "total",
                      "tip",
                      "type",
                      "productIDs",
                    ],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": " ",
                "className": "col-lg-8",
              },
              "ref": null,
              "rendered": " ",
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="btn btn-primary"
                onClick={[Function]}
              >
                Create scatter: (
                , 
                )
              </button>,
              <div />,
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Create scatter: (",
                  undefined,
                  ", ",
                  undefined,
                  ")",
                ],
                "className": "btn btn-primary",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                "Create scatter: (",
                undefined,
                ", ",
                undefined,
                ")",
              ],
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [],
              },
              "ref": null,
              "rendered": Array [],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
